### ISSUE TITLE
Load .leo.db files with sqlite

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#514]
Discussed [here ](https://groups.google.com/d/msg/leo-editor/ouoTVKGg5QA/ClLwGRcLBgAJ)and implemented in the [sqlite-format](https://github.com/leo-editor/leo-editor/tree/sqlite-format) branch

~This is highly experimental.~

**Comment by vitalije** [comment]
I found that leo/core/LeoPy.db is added to git.  It is a binary file, possible large one. Is it wise to keep it iside git?

**Comment by vitalije** [comment]
I have deleted leoPy.db from git index to save band-width when fetching from this branch. If it was wrong move I apology. It can be restored.

**Comment by edreamleo** [comment]
On Sat, Jul 1, 2017 at 2:22 PM, Виталије Милошевић <notifications@github.com
> wrote:

> I found that leo/core/LeoPy.db is added to git. It is a binary file,
> possible large one. Is it wise to keep it iside git?
>
​I think deleting it is best.

Edward

**Comment by vitalije** [comment]
For about two weeks of intense work on Leo development I have used sqlite-leo branch and db version of LeoPyRef outline. The only trouble I had was mysterious Leo reports that LeoPyRef.db file was changed outside of Leo and asking what to do. A db version of Leo document is actually never overwritten and this question is totally unnecessary. I have disabled it in some parts of Leo code, but it somehow continue to pass all my barriers and ask again and again. 

Apart from this smallish annoyance, I haven't noticed any problems in normal workflow. It seemed that sqlite-leo branch was ready for production.

However, the other day, I have encountered serious problems with switching branches, and hard to replicate strange behavior of 'refresh-from-disk'. I think until this bug is fully discovered, explained and fixed, this branch can't be announced as ready for production. So my next task is to try to dig it out and fix it thoroughly. 

I had very rough day with it. I tried to use git bisect to find out when this bug was introduced. I used git bisect earlier but this time it gave me new additional troubles. Maybe I don't understand how git works with branching and merging, but it seems that git bisecting part of code history, when there are a lot of merges (like in sqlite-leo branch now), jumps arbitrarily from one branch to the other. I couldn't believe myself what was happening. In some attempts during bisecting, Leo wouldn't launch at all and it wasn't easy to kill it either. Later I realized that it was Leo from master branch trying to open few mega large db file as plain text file. I still don't understand why git tried that code at all. I have updated git to latest version. In one moment it seemed to me that all recent commits in sqlite-leo were wrong. 

The solution I found was to delete my db file, delete all caches and open again xml version of LeoPyRef.leo and save it as new db file. Lesson learned was that it is better to keep different branches in different folders and avoid switching branches altogether.

My prime suspect is leoCacher, but I can't be sure. It could be also merging branch new-read into master and merging master afterwards in sqlite-leo. I tried to explore what has changed in new-read branch that could affect sqlite-leo but haven't find anything suspicious so far.

Using leoCacher when in sqlite-leo branch is a bit awkward. I tried to keep minimal changes in sqlite-leo to make it easier to merge later. So, I didn't touch leoCacher module at all. Now, I intend to create another cacher module and make decision which one to import with g.SQLITE switch. In essence db file is already cache, keeping cached versions of all nodes and also keeping hashes of all external files. I just need to check other uses of leoCacher code. 

I also plan to make similar parallel leoAtFile module that would use new lines property of VNode class. While doing it I intend to try to make these two modules truly reloadable, so that they can be tested using generative tests. (I learned about this technique in Clojure(script) world). Using this technique it is possible to generate smallest sequence of steps that lead to erroneous state. If it doesn't find such sequence in reasonable amount of time, code can be considered to be bug free with very high probability.

I will report here when I have something to show (or more probably in separate branch).

**Comment by edreamleo** [comment]
@vitalije What is the status of this?  Has this been merged into the trunk?

**Comment by edreamleo** [comment]
I am going to close this, because we have been using this code without problems for almost two week.

Many thanks, Vitalije, for this work.

**Comment by vitalije** [comment]
@edreamleo  No, it is hasn't been merged yet. The message you have quoted is about using sqlite for Leo cache. I've been using sqlite-leo version for all my development ever since it was published and had no problem with it so far. If you agree I will merge it now. The only thing that is missing is documentation of this new feature and some unresolved questions about preferences of the output file format. Leo has some hard-coded decisions about file extension if user doesn't explicitly enter file extension. Perhaps we can add a setting for this?

**Comment by edreamleo** [comment]
On Fri, Aug 11, 2017 at 4:01 AM, Виталије Милошевић <
notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo> No, it is hasn't been merged
> yet. The message you have quoted is about using sqlite for Leo cache. I've
> been using sqlite-leo version for all my development ever since it was
> published and had no problem with it so far. If you agree I will merge it
> now.
>
​Is this the sqlite-leo branch?​ Perhaps there should be a pull request.

> The only thing that is missing is documentation of this new feature and
> some unresolved questions about preferences of the output file format. Leo
> has some hard-coded decisions about file extension if user doesn't
> explicitly enter file extension. Perhaps we can add a setting for this?
>
Let's pretend I don't remember anything about this, which is mostly true
;-)  I'd like to see full documentation before I merge the PR.  It's find
to cut/paste from previous discussions.

Edward

**Comment by vitalije** [comment]
@edreamleo  here is first draft of documentation:

# New file format for Leo documents
Leo now supports new `.db` file format. It is actually SQLite database
file. In current version all data are stored in two tables:

   1. vnodes which contains all vnodes from outline stored one node per row.
   2. extra_infos which contains (key, value) pairs for additional informations
      like currently selected position, screen position and dimension of Leo
      application window, and such.

Plugins can create and use additional tables.

There is also new setting `@string default_leo_extension = .leo`.
When saving to new file if user doesn't provide extension, this
default file extension will be appended to filename. Leo will honor
the extension explicitly set by user.

When editing `.db` files `c.sqlite_connection` is `sqlite3.Connection` instance connected to `.db` file.

In current implementation, Leo operates data base just when opening and saving Leo document. On every open/save command it retrieves/stores all nodes. This can be changed to update nodes on every change and to retrieve nodes on demand. But for now this all-in/all-out strategy is satisfying.

I have no more ideas what to say about this feature. Please ask me if there is something that needs more explanation and correct my English. 
I have updated branch sqlite-leo to match master HEAD. It is automatically merge-able. 

Vitalije

**Comment by edreamleo** [comment]
On Fri, Aug 11, 2017 at 9:14 AM, Виталије Милошевић <
notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo> here is first draft of
> documentation:
> New file format for Leo documents
>
> Leo now supports new .db file format. It is actually SQLite database
> file. In current version all data are stored in two tables:
>
>    1. vnodes which contains all vnodes from outline stored one node per
>    row.
>    2. extra_infos which contains (key, value) pairs for additional
>    informations
>    like currently selected position, screen position and dimension of Leo
>    application window, and such.
>
> Plugins can create and use additional tables.
>
> There is also new setting @string default_leo_extension = .leo.
> When saving to new file if user doesn't provide extension, this
> default file extension will be appended to filename. Leo will honor
> the extension explicitly set by user.
>
> When editing .db files c.sqlite_connection is sqlite3.Connection instance
> connected to .db file.
>
> In current implementation, Leo operates data base just when opening and
> saving Leo document. On every open/save command it retrieves/stores all
> nodes. This can be changed to update nodes on every change and to retrieve
> nodes on demand. But for now this all-in/all-out strategy is satisfying.
>
​This looks good to me.  Would it be a bother to use .leo.db instead of
.db?  I think this would be more informative.

Edward

**Comment by vitalije** [comment]
`.db` includes `.leo.db`. We can suggest that it is preferred or conventional extension, but we should keep in mind that lot of functions that deal with paths and filenames are not designed for extensions that contain dot. For example in python: `os.path.splitext('test.leo.db')` gives `('test.leo', '.db')`.

Also on Windows (at least in the age of Windows XP) only first three letters of extension were important and the rest didn't matter. I don't know if it is still true in newer versions of Windows. On linux, however, extension is much less important and mime type of file is deduced by its content.

I don't know what Qt expect and how it will treat file filters with extensions containing dot.

In short I have nothing against `.leo.db` extension, but IMO it will be enough to keep in documentation `.leo.db` as official extension and code can remain as it is and it will open both `.db` and `.leo.db` files. Perhaps only where Leo adds automatically default extension, we can have `.leo.db` to be added by default instead of `.db` as it is now.

**Comment by edreamleo** [comment]
On Fri, Aug 11, 2017 at 1:17 PM, Виталије Милошевић <
notifications@github.com> wrote:

> .db includes .leo.db. We can suggest that it is preferred or conventional
> extension, but we should keep in mind that lot of functions that deal with
> paths and filenames are not designed for extensions that contain dot.
>
​Ok then. As I understand it, this work should only affect people who set​
@string default_leo_extension = .db. That being so, it should be safe to
merge this work into master.  Please do so, unless I have misunderstood
something.

Edward

**Comment by vitalije** [comment]
Merged in  5c50fd0.
All unit tests pass. I have first merged master in sqlite-leo and check it once again. When I merged in master branch, it didn't offer me to write different commit message and  5c50fd0  message says merged master into sqlite-leo, but it really should say:  `merged sqlite-leo in master.`

**Comment by edreamleo** [comment]
I am going to reopen this.  I am not able to load
`#@+node:vitalije.20170701155512.1: *6* at.checkExternalFileAgainstDb` into my copy of leoPy.leo.

**Comment by vitalije** [comment]
Hm, I haven't touch LeoPyRef.leo, perhaps that is problem. I have now open LeoPyRef.leo, executed read-at-file nodes on `Code` node and saved LeoPyRef.leo. 
Please try b3863324e

**Comment by vitalije** [comment]
Also executing `read-at-file-nodes` should fix. I am not sure what did happen? Have you been able to open LeoPyRef.leo at all?

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 7:08 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Also executing read-at-file-nodes should fix. I am not sure what did
> happen? Have you been able to open LeoPyRef.leo at all?
>
​The problem is in loading leoPy.leo.  I am investigating by adding traces
(in an external editor).  Imo, it's important to understand exactly what is
happening before attempting any changes or workarounds.  This is too good
an opportunity to miss.

Edward

**Comment by edreamleo** [comment]
Hah.  Removing the `@c` comment preceding the line:
`def checkExternalFileAgainstDb(self, root):`
causes the node to appear normally.

Could this be an ancient bug that just happened to show up now?  It's possible!  Afaik, there are no other such comments in Leo's core.

**Comment by edreamleo** [comment]
I'll be responsible for this investigation.  The first step will be to check out a rev before the merge happened.  This has the smell of something unrelated to the sqlite-leo branch.

**Comment by vitalije** [comment]
Ok. I am glad that you have found it. However, I am curios why it didn't hit me? Perhaps we have different plugins enabled? Is it maybe that at-c looks like python decorator? Will it load correctly if after at-c line you put an empty line or just `pass`?

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 7:44 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Ok. I am glad that you have found it. However, I am curios why it didn't
> hit me?
>
​I too am wondering.​

> Perhaps we have different plugins enabled? Is it maybe that at-c looks
> like python decorator? Will it load correctly if after at-c line you put an
> empty line or just pass?
>
​Good question. I'll check.​

Edward

**Comment by edreamleo** [comment]
> On Sat, Aug 12, 2017 at 7:44 AM, Виталије Милошевић <
> notifications@github.com> wrote:
>
>> Ok. I am glad that you have found it. However, I am curios why it didn't
>> hit me?
>>
> ​I too am wondering.​
>
>> Perhaps we have different plugins enabled? Is it maybe that at-c looks
>> like python decorator? Will it load correctly if after at-c line you put an
>> empty line or just pass?
>>
> ​Good question. I'll check.​
>

​Inserting either a blank line or a 'pass' statement fixes the problem.  It
looks like Leo's atFile write code is somehow confused. To repeat, this
looks like a bug in leoAtFile.py.

Could it be that everything was always cached when you were developing the
sqlite-leo branch?

Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 8:01 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

​Inserting either a blank line or a 'pass' statement fixes the problem.  It
> looks like Leo's atFile write code is somehow confused. To repeat, this
> looks like a bug in leoAtFile.py.
>

​On second thought, I'm not so sure. Leo's write code doesn't know anything
about python, or any other language...

I'll do some tests in an hour or so...

Edward

**Comment by vitalije** [comment]
I have made first edits using master branch, and as soon as sqlite-leo branch was functional I have used it exclusively. It caches all nodes in db and not in cacher instance. Most probably this comment I added later using sqlite-leo branch. 

I have just deleted cache and reopen LeoPyRef.leo without problem?

Though, I think maybe I need to try to open LeoPyRef.leo from earlier version that has not this particular node. I 'll try this and report

**Comment by vitalije** [comment]
I have just deleted cache again, checked out LeoPyRef.leo an earlier version (before sqlite-leo merge) and open it using most current version. It has read normally all at-file nodes.

**Comment by vitalije** [comment]
Could it be that your LeoPy.leo uses older version of Leo file format? I have found following in Leo documentation:
### New in 4.0:
  *  <v> elements corresponding to @file nodes now contain tnodeList attributes. The tnodeList attribute allows Leo to recreate the order in which nodes should appear in the outline. The tnodeList attribute is a list of gnx’s: global node indices. See Format of external files (4.x) for the format of gnx’s.

I can see `tnodeList` throughout leoAtFile.py, but I can't see it in LeoPyRef.leo. Is there any attribute `tnodeList` in your LeoPy.leo?

What if your LeoPy.leo has list of gnxes that it expects and there is newly added node with gnx that is not in tnodeList? I am not sure what tnodeList is and how is it meant to work. I am just guessing relying on the above quote from the documentation.

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 8:51 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Could it be that your LeoPy.leo uses older version of Leo file format?
>
I doubt it.  leoPy.leo contains:

<leo_header file_format="2" tnodes="0" max_tnode_index="0"
clone_windows="0"/>

> I can see tnodeList throughout leoAtFile.py, but I can't see it in
> LeoPyRef.leo. Is there any attribute tnodeList in your LeoPy.leo?
>
​A search reveals no tnodeList attribute in leoPy.leo.

> What if your LeoPy.leo has list of gnxes that it expects and there is
> newly added node with gnx that is not in tnodeList? I am not sure what
> tnodeList is and how is it meant to work. I am just guessing relying on the
> above quote from the documentation.
>
​The tnodeList is ancient history, iirc.


I just checked out an older version of Leo.  There are no problems handling
@ comments.


It might be best to undo the merge of the sqlite-leo branch.  I can then do
the testing I should have done earlier in the sqlite-leo branch itself.


Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 8:51 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Could it be that your LeoPy.leo uses older version of Leo file format?
>
​Here is another clue.  The following are being deleted when I load
leoPy.leo.  I don't even have to save leoPy.leo for this to happen:

​*leoAtFile.py*:
#@+node:vitalije.20170701155512.1: *6* at.checkExternalFileAgainstDb

*leoCommands.py*:
#@+node:ekr.20170808014610.1: *5* c.enable/disable_redraw (New in Leo 5.6)

#@+node:ekr.20170808005711.1: *5* c.redraw_later

*cursesGui2.py*:
#@+node:ekr.20170712145632.2: *5* CGui.createFindDialog

*leoTest.py*:
A minor whitespace change, perhaps involving line endings.

This repeatable.  git checkout leo, delete contents of .leo.db, reload
leoPy.leo.  *Don't* save leoPy.leo.  Now git status reports the above four
files as changed, and git diff shows the missing nodes.

The conjecture that @ comments are the culprit appears to be false.

Edward

**Comment by vitalije** [comment]
This must be something related with cache. When I open your leoPy.leo for the first time (without cache) it loads all files perfectly and even when I force write of all external files, all of them are reported as unchanged.

I suspect maybe some line-ending difference may be in effect. I will investigate further. 
I have copied the cache.sqlite left after first successful load and in cache is stored version of leoAtFile.py that contains missing node.

**Comment by vitalije** [comment]
I have found that there is an error in leoPy.leo. There are two nodes with the same gnx and yet they are not clones:
```
leoPy.leo:393: <v t="ekr.20041005105605.21"><vh>at.read &amp; helpers</vh>
leoPy.leo:2310: <t tx="ekr.20041005105605.21">def read(self, root, importFileName=None,
```
Don't know how it happen, but that is clearly the cause that parent node of missing node `at.checkExternalFileAgainstDb` is considered to be already created that is why it skips its children.

Probably, similar error is in other problematic files.

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 2:49 PM, Виталије Милошевић <
notifications@github.com> wrote:

> I have found that there is an error in leoPy.leo. There are two nodes with
> the same gnx and yet they are not clones:
>
> leoPy.leo:393: <v t="ekr.20041005105605.21"><vh>at.read &amp; helpers</vh>
> leoPy.leo:2310: <t tx="ekr.20041005105605.21">def read(self, root, importFileName=None,
>
> Don't know how it happen, but that is clearly the cause that parent node
> of missing node at.checkExternalFileAgainstDb is considered to be already
> created that is why it skips its children.
>
​Excellent sleuthing.  It's good that the sqlite-leo branch does not appear
to be at fault.

It will be important not simply to ignore this bad file. Rather I'll see if
I can discover how this happened, and whether there is a way of preventing
it from happening in future.

Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 3:00 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

>
> On Sat, Aug 12, 2017 at 2:49 PM, Виталије Милошевић <
> notifications@github.com> wrote:
>
>> I have found that there is an error in leoPy.leo. There are two nodes
>> with the same gnx and yet they are not clones:
>>
>> leoPy.leo:393: <v t="ekr.20041005105605.21"><vh>at.read &amp; helpers</vh>
>> leoPy.leo:2310: <t tx="ekr.20041005105605.21">def read(self, root, importFileName=None,
>>
>> ​On second look, ​everything appears correct.

The first line is the <v> element.  The second line is the corresponding<t>
element.  Although Leo no longer uses tnodes, the file format still uses
both <t> and <v> elements.

Don't know how it happen, but that is clearly the cause that parent node of
>> missing node at.checkExternalFileAgainstDb is considered to be already
>> created that is why it skips its children.
>>
> ​I'll investigate thoroughly.

Edward

**Comment by vitalije** [comment]
I can't say that this is something that is likely to happen, but it could never happen if we edited the same reference file. I am referring to #538 and #540. When we both edit our own private copies and usually don't touch reference file at all, our files can go out of sync with the reference file and what is more important we don't have a history of our private copies and we can't figure out when has this happened. 

When reading external files Leo does update body of changed clones, but it doesn't check if cloned nodes have same list of children. It checks only those children that has already been in the cache list not children of vnode from gnxDict.
```python
for grand_child in grand_children:
    self.checkForChangedNodes(grand_child, fileName, child_v)
```
lines 190-191 from leoCache.py
`grand_children` is list of nodes from cache and `child_v` is node from xml (or previously read external files). It is possible that `child_v` has more children than `grand_children` list has.

**Comment by vitalije** [comment]
No, I think it is opposite case. When cache has more children than (and that is what happened here, after first successful open cache has all nodes that are present in external file), then `checkForChangedNodes`, unless `trace = True`, will silently return (`else` branch) without any warning.

**Comment by vitalije** [comment]
I have found where is the error. Fix is in trunk. f5918a47e

**Comment by vitalije** [comment]
if  node was clone `createOutlineFromCacheList` was not going in depth. Now, it is outside `if` statement. If it is clone then perform checks for changed clones, but in any case descend down tree.

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 3:45 PM, Виталије Милошевић <
notifications@github.com> wrote:

> I have found where is the error. Fix is in trunk. f5918a4
> <https://github.com/leo-editor/leo-editor/commit/f5918a47e6b90c3a9b84093b3be8e74c52a54846>
>
​Oh my.  Removing that else clause looks like a major change.

If your fix is correct, it will have all kinds of happy effects.

How did you find this?

Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 3:49 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

> I have found where is the error. Fix is in trunk. f5918a4
>> <https://github.com/leo-editor/leo-editor/commit/f5918a47e6b90c3a9b84093b3be8e74c52a54846>
>>
> ​Oh my.  Removing that else clause looks like a major change.
>

​Alas, removing the else clause​ creates all kinds of incorrect clones.
When re-writing files, pyflakes warns about this.  For example, when
rewriting leoAtFile.py, the first complaint is:

leoAtFile.py:645: redefinition of unused 'createResurrectedNodesNode' from
line 600

Please revert f5918a.  It's not correct.

I'm not sure what the sqlite-leo branch is doing, but let's assume for now
that leoCache.py is correct.

Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 3:58 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

> On Sat, Aug 12, 2017 at 3:49 PM, Edward K. Ream <edreamleo@gmail.com>
> wrote:
>
>> I have found where is the error. Fix is in trunk. f5918a4
>>> <https://github.com/leo-editor/leo-editor/commit/f5918a47e6b90c3a9b84093b3be8e74c52a54846>
>>>
>> ​Oh my.  Removing that else clause looks like a major change.
>>
>
> ​Alas, removing the else clause​ creates all kinds of incorrect clones.
> When re-writing files, pyflakes warns about this.  For example, when
> rewriting leoAtFile.py, the first complaint is:
>

To clarify, these extra clones may well depend on what clones already exist
in leoPy.leo.

Edward

**Comment by vitalije** [comment]
I have reverted the last change, and I was not testing it any further then trying to open leoPy.leo and checking to see if there is node that was missing. Sorry about that.

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 4:05 PM, Виталије Милошевић <
notifications@github.com> wrote:

> I have reverted the last change, and I was not testing it any further then
> trying to open leoPy.leo and checking to see if there is node that was
> missing. Sorry about that.
>

​No problem.  This is a difficult project.

Edward

**Comment by vitalije** [comment]
Another attempt at 2c497cd60.
I believe it is real fix now. All unit test pass.

**Comment by vitalije** [comment]
And leoPy.leo opens correctly.

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 4:36 PM, Виталије Милошевић <
notifications@github.com> wrote:

> And leoPy.leo opens correctly.
>
​This is looking better.  The "acid test" still writes some files, but this
appears harmless.

Git reports:
    modified:   leo/commands/editFileCommands.py
    modified:   leo/core/leoCommands.py
    modified:   leo/core/leoTest.py

This may be due to a mismatch between leoPy.leo and ​LeoPyRef.leo.  All
unit tests pass with the changed files.

I have to go now, but I'll commit any necessary "cleanup" changes later
tonight.

Edward

**Comment by edreamleo** [comment]
On Sat, Aug 12, 2017 at 5:08 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

I have to go now, but I'll commit any necessary "cleanup" changes later
> tonight.
>

​Done at 88b5814.  Everything appears to work for me, but there is no
guarantee that others may have problems.

Edward

**Comment by xgid** [comment]
A minor detail I've just noticed: the [README.MD](https://github.com/leo-editor/leo-editor/blob/master/README.MD) currently in master seems to be the one from the sqlite-leo branch, not the official Leo one.

**Comment by vitalije** [comment]
@xgid good catch.

**Comment by vitalije** [comment]
@edreamleo should we close this?

**Comment by edreamleo** [comment]
On Tue, Aug 15, 2017 at 1:11 AM, Виталије Милошевић <
notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo> should we close this?
>
​Not yet.  I'd like to investigate (and fix) the "reading" messages.

Also, with Python 2 only, I get this crash during startup. It causes Leo to
exit:

Traceback (most recent call last):
  File "c:\leo.repo\leo-editor\launchLeo.py", line 8, in <module>
    leo.core.runLeo.run()
  File "c:\leo.repo\leo-editor\leo\core\runLeo.py", line 63, in run
    g.app.loadManager.load(fileName, pymacs)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2109, in load
    ok = lm.doPostPluginsInit()
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2711, in
doPostPluginsInit
    c = lm.loadLocalFile(fn, gui=g.app.gui, old_c=None)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2945, in
loadLocalFile
    previousSettings = lm.getPreviousSettings(fn)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 1863, in
getPreviousSettings
    c = lm.openSettingsFile(fn)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2028, in
openSettingsFile
    readAtFileNodesFlag=False, silent=True)
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 841, in
openLeoFile
    silent=silent,
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 672, in
getLeoFile
    p = recoveryNode or c.p or c.lastTopLevel()
  File "c:\leo.repo\leo-editor\leo\core\leoCommands.py", line 6629, in
lastTopLevel
    while p.hasNext():
AttributeError: 'NoneType' object has no attribute 'hasNext'

Edward

**Comment by vitalije** [comment]
Regarding messages try  1f10c61  I believe line endings are different and that is why no caching was done.

**Comment by edreamleo** [comment]
On Tue, Aug 15, 2017 at 8:41 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Regarding messages try 1f10c61
> <https://github.com/leo-editor/leo-editor/commit/1f10c6126b61d8e13bcbf7ed233bdccff30afe4f>
> I believe line endings are different and that is why no caching was done.
>
​Looks good with Python 2.  With Python 3 Leo exits during startup​:

reading settings in c:\leo.repo\leo-editor\leo\core\leoPy.db
Traceback (most recent call last):
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 657, in
getLeoFile
    ok = fc.getLeoFileHelper(theFile, fileName, silent)
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 720, in
getLeoFileHelper
    reassignIndices=False,
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 1291, in
readSaxFile
    return fc.retrieveVnodesFromDb(theFile) # or fc.initNewDb(theFile) -
not implemented yet
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 1410, in
retrieveVnodesFromDb
    assert rootChildren, 'there should be at least one top level node'
AssertionError: there should be at least one top level node

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "launchLeo.py", line 8, in <module>
    leo.core.runLeo.run()
  File "c:\leo.repo\leo-editor\leo\core\runLeo.py", line 63, in run
    g.app.loadManager.load(fileName, pymacs)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2109, in load
    ok = lm.doPostPluginsInit()
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2711, in
doPostPluginsInit
    c = lm.loadLocalFile(fn, gui=g.app.gui, old_c=None)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2945, in
loadLocalFile
    previousSettings = lm.getPreviousSettings(fn)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 1863, in
getPreviousSettings
    c = lm.openSettingsFile(fn)
  File "c:\leo.repo\leo-editor\leo\core\leoApp.py", line 2028, in
openSettingsFile
    readAtFileNodesFlag=False, silent=True)
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 841, in
openLeoFile
    silent=silent,
  File "c:\leo.repo\leo-editor\leo\core\leoFileCommands.py", line 672, in
getLeoFile
    p = recoveryNode or c.p or c.lastTopLevel()
  File "c:\leo.repo\leo-editor\leo\core\leoCommands.py", line 6629, in
lastTopLevel
    while p.hasNext():
AttributeError: 'NoneType' object has no attribute 'hasNext'

c:\leo.repo\leo-editor>

Edward

**Comment by vitalije** [comment]
@edreamleo  this crash in Python2, what file were you trying to open `.leo.db` or `.leo`. According to the trace back it happen while opening settings file and the reason is most probably that file doesn't contain any node so the outline is empty and `lastTopLevel` doesn't exist.

I have just tried to open non existent `.db` file and Leo crashed. There is comment saying that this still isn't implemented. Perhaps it would be simplest to init empty db with just one `topLevelNode` as it happen when opening non existent `.leo` file

**Comment by vitalije** [comment]
Ok. I am working to solve both issues non caching and opening empty `.db` file.

**Comment by edreamleo** [comment]
On Tue, Aug 15, 2017 at 8:49 AM, Виталије Милошевић <
notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo> this crash in Python2, what
> file were you trying to open .leo.db or .leo. According to the trace back
> it happen while opening settings file and the reason is most probably that
> file doesn't contain any node so the outline is empty and lastTopLevel
> doesn't exist.
>
> I have just tried to open non existent .db file and Leo crashed. There is
> comment saying that this still isn't implemented. Perhaps it would be
> simplest to init empty db with just one topLevelNode as it happen when
> opening non existent .leo file
>
​Sounds reasonable to me. But leoPy.db definitely exists, and it looks like
a .db file.

Edward

**Comment by vitalije** [comment]
Done at b603634.

Leo now avoids to throw exceptions in unexpected conditions while reading database. I have also switched protocol for pickle.dumps to enable sharing same `.leo.db` file while using both python2 and python3. (Before files written with python3 could not be opened with python2 because it was used highest protocol for pickle).

Opening empty `.leo.db` (or non existent) follows the same principle as when opening empty `.leo` file. It creates one top level node.

Caching used to be implemented by directly reading bytes from external file. However, on Windows bytes read that way contain `'\r\n'` combinantion, while when checking if cache is valid it was used normalized version of file content with `'\r\n'` replaced by `'\n'` and that was the reason why cache never was valid.

I hope this concludes #514. 

If any problem should arise, perhaps we can open new issue. This one has become quite long.

**Comment by edreamleo** [comment]
ac991c fixes the last nits.
